src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,867 @@
+
+
+                resource_metrics.scope_metrics.funcs.encode =
+                    &encode_scope_metrics;
+                resource_metrics.scope_metrics.arg = &scopeMetrics;
+
+                metrics_data.resource_metrics.funcs.encode =
+                    &encode_resource_metrics;
+                metrics_data.resource_metrics.arg = &resource_metrics;
+        }
+
+        status = pb_get_encoded_size(
+            &message_size, opentelemetry_proto_metrics_v1_MetricsData_fields,
+            &metrics_data);
+        if (!status) {
+                rd_kafka_dbg(rk, TELEMETRY, "RD_KAFKA_TELEMETRY_METRICS_INFO",
+                             "Failed to get encoded size");
+                free_metrics(metrics, metric_names, data_points,
+                             datapoint_attributes_key_values,
+                             total_metrics_count);
+                free_resource_attributes(resource_attributes_key_values,
+                                         resource_attributes_struct,
+                                         resource_attributes_count);
+                return NULL;
+        }
+
+        buffer = rd_malloc(message_size);
+        if (buffer == NULL) {
+                rd_kafka_dbg(rk, TELEMETRY, "RD_KAFKA_TELEMETRY_METRICS_INFO",
+                             "Failed to allocate memory for buffer");
+                free_metrics(metrics, metric_names, data_points,
+                             datapoint_attributes_key_values,
+                             total_metrics_count);
+                free_resource_attributes(resource_attributes_key_values,
+                                         resource_attributes_struct,
+                                         resource_attributes_count);
+
+                return NULL;
+        }
+
+        stream = pb_ostream_from_buffer(buffer, message_size);
+        status = pb_encode(&stream,
+                           opentelemetry_proto_metrics_v1_MetricsData_fields,
+                           &metrics_data);
+
+        if (!status) {
+                rd_kafka_dbg(rk, TELEMETRY, "RD_KAFKA_TELEMETRY_METRICS_INFO",
+                             "Encoding failed: %s", PB_GET_ERROR(&stream));
+                rd_free(buffer);
+                free_metrics(metrics, metric_names, data_points,
+                             datapoint_attributes_key_values,
+                             total_metrics_count);
+                free_resource_attributes(resource_attributes_key_values,
+                                         resource_attributes_struct,
+                                         resource_attributes_count);
+
+                return NULL;
+        }
+        free_metrics(metrics, metric_names, data_points,
+                     datapoint_attributes_key_values, total_metrics_count);
+        free_resource_attributes(resource_attributes_key_values,
+                                 resource_attributes_struct,
+                                 resource_attributes_count);
```

Comment:
Move this after `reset_historical_metrics` with a `rd_kafka_rdunlock`
```suggestion
        free_metrics(metrics, metric_names, data_points,
                     datapoint_attributes_key_values, total_metrics_count);
        free_resource_attributes(resource_attributes_key_values,
                                 resource_attributes_struct,
                                 resource_attributes_count);
        rd_kafka_rdunlock(rk);
        
        return rbuf;
fail:
        free_metrics(metrics, metric_names, data_points,
                     datapoint_attributes_key_values, total_metrics_count);
        free_resource_attributes(resource_attributes_key_values,
                                 resource_attributes_struct,
                                 resource_attributes_count);

        rd_kafka_rdunlock(rk);
        return NULL;
```

Reply:
Updated